frontend/src/v5/ui/routes/viewer/viewer.tsx

```diff
@@ -17,21 +17,51 @@
 
 import { ViewerGui } from '@/v4/routes/viewerGui';
 import { useParams } from 'react-router-dom';
+import { ContainersHooksSelectors } from '@/v5/services/selectorsHooks/containersSelectors.hooks';
+import { FederationsHooksSelectors } from '@/v5/services/selectorsHooks/federationsSelectors.hooks';
+import { InvalidContainerOverlay, InvalidFederationOverlay } from './invalidViewerOverlay';
 import { ViewerParams } from '../routes.constants';
+import { CheckLatestRevisionReadiness } from './checkLatestRevisionReadiness/checkLatestRevisionReadiness.container';
 import { useContainersData } from '../dashboard/projects/containers/containers.hooks';
 import { useFederationsData } from '../dashboard/projects/federations/federations.hooks';
 
 export const Viewer = () => {
 	const { teamspace, containerOrFederation, revision } = useParams<ViewerParams>();
+
+	useContainersData();
+	useFederationsData();
+
+	const areStatsPending = FederationsHooksSelectors.selectAreStatsPending();
+	const isListPending = FederationsHooksSelectors.selectIsListPending();
+	const isLoading = areStatsPending || isListPending;
+
+	const selectedContainer = ContainersHooksSelectors.selectContainerById(containerOrFederation);
+	const selectedFederation = FederationsHooksSelectors.selectFederationById(containerOrFederation);
+	const federationsContainers = FederationsHooksSelectors.selectContainersByFederationId(containerOrFederation);
+	const federationIsEmpty = selectedFederation?.containers?.length === 0
+		|| federationsContainers.every((container) => container?.revisionsCount === 0);
+
 	const v4Match = {
 		params: {
 			model: containerOrFederation,
 			teamspace,
 			revision,
 		} };
 
-	useFederationsData();
-	useContainersData();
+	const handleEmptyViewer = () => {
+		if (selectedContainer?.revisionsCount === 0) {
+			return <InvalidContainerOverlay status={selectedContainer.status} />;
+		}
+		if (selectedFederation && federationIsEmpty) {
+			return <InvalidFederationOverlay containers={federationsContainers} />;
+		}
+		return (
+			<>
+				<CheckLatestRevisionReadiness />
+				<ViewerGui match={v4Match} />;
+			</>
+		);
+	};
 
-	return <ViewerGui match={v4Match} />;
+	return isLoading ? (<></>) : handleEmptyViewer();
```

Comment:
In general I prefer not to have explicit functions to return a component (I  know that there can be functional components but the syntaxes is different and makes you think about a component ). This feels a bit imperative. 

I feel that is more readable this way; its simple to see that there are cases that replace the whole page and the condition that trigger those, and it states that the normal condition of the component is to return a ViewerGui. (this would start in line 44, github would let me select the whole block)


But would like to hear your thoughts on this.

```suggestion

	if (isLoading) {
		return (<></>);
	}

	if (selectedContainer?.revisionsCount === 0) {
		return <InvalidContainerOverlay status={selectedContainer.status} />;
	}

	if (selectedFederation && federationIsEmpty) {
		return <InvalidFederationOverlay containers={federationsContainers} />;
	}

	const v4Match = {
		params: {
			model: containerOrFederation,
			teamspace,
			revision,
		},
	};

	return (
		<>
			<CheckLatestRevisionReadiness />
			<ViewerGui match={v4Match} />;
		</>
	);
```

Reply:
I think I used the ternary operator because I thought it'd make it clearer that you only handle the empty viewers once everything is loaded, but your method looks a lot clearer, so I'll do that. I just like ternary operators ¯\_(ツ)_/¯